src/lib/markets/graph.ts

```diff
@@ -983,3 +983,129 @@ export const getPriceAtVersion = async ({
 
   return res.marketVersionPrice?.price ?? 0n
 }
+
+export async function fetchTradeHistory({
+  graphClient,
+  address,
+  first,
+  offset,
+}: {
+  graphClient: GraphQLClient
+  address: Address
+  first: number
+  offset: number
+}) {
+  const tradeHistoryQuery = gql(`
+   query fetchTradeHistory($account: Bytes!, $first: Int!, $offset: Int!) {
+    accountPositionProcesseds(
+      where: { account: $account },
+      orderBy: toOracleVersion,
+      orderDirection: desc,
+      first: $first,
+      skip: $offset,
+    ) {
+      accumulationResult_collateralAmount, accumulationResult_keeper, accumulationResult_positionFee, priceImpactFee
+        accumulatedPnl, accumulatedFunding, accumulatedInterest, accumulatedMakerPositionFee, accumulatedValue
+        side, size, fromOracleVersion, toOracleVersion, toVersionPrice, toVersionValid, collateral, blockNumber,
+        market
+      update {
+        version, collateral, newMaker, newLong, newShort, valid, transactionHash, price, priceImpactFee,
+        localPositionId, globalPositionId, market, account, blockNumber, blockTimestamp, protect, interfaceFee, orderFee, side, delta
+      }
+    }
+   }
+  `)
+
+  const { accountPositionProcesseds } = await graphClient.request(tradeHistoryQuery, {
+    account: address,
+    first,
+    offset,
+  })
+
+  const entities = accountPositionProcesseds.reduce(
+    (acc, entity) => {
+      if (!acc[entity.market]) acc[entity.market] = []
+      acc[entity.market].push(entity)
+      return acc
+    },
+    {} as Record<string, typeof accountPositionProcesseds>,
+  )
+
+  type Update = (typeof accountPositionProcesseds)[number]['update'] & {
+    accountPositionProcesseds: typeof accountPositionProcesseds
+  }
+
+  const trades = Object.entries(entities).reduce(
+    (acc, [market, entities]) => {
+      const updates: Update[] = []
+
+      entities.forEach((entity) => {
+        if (
+          entity.update &&
+          entity.update.side !== 'none' &&
+          entity.update.delta !== '0' &&
+          entity.update.collateral !== '0'
+        ) {
+          updates.push({ ...entity.update, accountPositionProcesseds: [entity] })
+        } else {
+          const update = updates[updates.length - 1]
+          if (update) {
```

Comment:
If the update doesn't exist we still need to add this to the next available update.

So I think we do something like `const update = updates.get(-1) ?? entities.find(e => <if statement above>)`